File: backend/agents/base-agent.ts
/**
 * Base Agent Class
 * Provides common functionality for all Azure AI Agents
 *
 * Note: Uses type casting in some places due to Azure AI SDK beta type inconsistencies.
 * The SDK API is functional at runtime even when TypeScript types don't match exactly.
 */

import type {
  AgentsClient,
  Agent,
  AgentThread,
  ThreadMessage,
  ThreadRun,
  ToolOutput,
  FunctionToolDefinition,
  RequiredToolCall,
} from '@azure/ai-agents';
import { getAgentsClient, getModelDeploymentName, buildAgentName } from './azure-client.js';
import { AgentConfig, AGENT_CONFIGS, AgentName } from './agent-config.js';

export interface ChatMessage {
  role: 'system' | 'user' | 'assistant';
  content: string;
}

export interface AgentResponse {
  success: boolean;
  message: string;
  metadata?: Record<string, any>;
  toolCalls?: Array<{
    name: string;
    arguments: Record<string, any>;
    result: any;
  }>;
}

export interface StreamEvent {
  type: 'status' | 'tool_call' | 'tool_result' | 'thinking' | 'complete' | 'error';
  data: Record<string, any>;
}

export type ToolHandler = (args: any) => Promise<any>;

/**
 * Abstract base class for Azure AI Agents
 */
export abstract class BaseAgent {
  protected agentName: AgentName;
  protected config: AgentConfig;
  protected agent: Agent | null = null;
  protected client: AgentsClient | null = null;
  protected toolHandlers: Map<string, ToolHandler> = new Map();

  constructor(agentName: AgentName) {
    this.agentName = agentName;
    this.config = AGENT_CONFIGS[agentName];

    if (!this.config) {
      throw new Error(`Unknown agent: ${agentName}`);
    }

    // Register tool handlers (to be overridden by subclasses)
    this.registerToolHandlers();
  }

  /**
   * Register tool handlers - to be overridden by subclasses
   */
  protected abstract registerToolHandlers(): void;

  /**
   * Initialize the agent - creates or retrieves from Azure
   */
  async initialize(): Promise<void> {
    console.log(`[${this.agentName}] Initializing agent...`);

    try {
      this.client = await getAgentsClient();
      const fullAgentName = buildAgentName(this.agentName);

      // Try to find existing agent
      const agents = this.client.listAgents();
      let existingAgent: Agent | undefined;

      for await (const agent of agents) {
        if (agent.name === fullAgentName) {
          existingAgent = agent;
          break;
        }
      }

      if (existingAgent) {
        console.log(`[${this.agentName}] Found existing agent: ${existingAgent.id}`);
        this.agent = existingAgent;
      } else {
        // Create new agent
        console.log(`[${this.agentName}] Creating new agent...`);
        this.agent = await this.client.createAgent(getModelDeploymentName(), {
          name: fullAgentName,
          instructions: this.config.instructions,
          tools: this.config.tools,
        });
        console.log(`[${this.agentName}] Created agent: ${this.agent.id}`);
      }
    } catch (error) {
      console.error(`[${this.agentName}] Failed to initialize:`, error);
      throw error;
    }
  }

  /**
   * Ensure agent is initialized
   */
  protected async ensureInitialized(): Promise<void> {
    if (!this.agent || !this.client) {
      await this.initialize();
    }
  }

  /**
   * Non-streaming chat - for backward compatibility
   */
  async chat(
    messages: ChatMessage[],
    context?: Record<string, any>
  ): Promise<AgentResponse> {
    await this.ensureInitialized();

    if (!this.client || !this.agent) {
      throw new Error('Agent not initialized');
    }

    const toolCalls: AgentResponse['toolCalls'] = [];

    try {
      // Create a thread
      const thread = await this.client.threads.create();

      // Add messages to thread
      for (const msg of messages) {
        if (msg.role !== 'system') {
          await this.client.messages.create(
            thread.id,
            msg.role === 'user' ? 'user' : 'assistant',
            msg.content
          );
        }
      }

      // Add context as a user message if provided
      if (context) {
        await this.client.messages.create(
          thread.id,
          'user',
          `[CONTEXT]: ${JSON.stringify(context)}`
        );
      }

      // Run the agent - using createAndPoll for simpler handling
      const poller = this.client.runs.createAndPoll(thread.id, this.agent.id);
      let run = await poller.pollUntilDone();

      // Poll for completion with tool call handling
      const maxIterations = 20;
      let iteration = 0;

      while (run.status === 'requires_action' && iteration < maxIterations) {
        // Handle required actions (tool calls)
        // Using type assertion due to SDK beta type inconsistencies
        const requiredAction = run.requiredAction as any;
        if (requiredAction?.submitToolOutputs?.toolCalls) {
          const toolOutputs = await this.handleToolCalls(
            requiredAction.submitToolOutputs.toolCalls,
            toolCalls
          );

          const submitResponse = this.client.runs.submitToolOutputs(
            thread.id,
            run.id,
            toolOutputs
          ) as any;

          // Wait for run to complete after submitting tool outputs
          run = await submitResponse.finalRun();
        } else {
          await new Promise(resolve => setTimeout(resolve, 1000));
          run = await this.client.runs.get(thread.id, run.id);
        }

        iteration++;
      }

      if (iteration >= maxIterations) {
        throw new Error('Agent run exceeded maximum iterations');
      }

      // Check for errors
      if (run.status === 'failed') {
        throw new Error(`Agent run failed: ${run.lastError?.message || 'Unknown error'}`);
      }

      // Get the response messages
      const messagesResponse = this.client.messages.list(thread.id);
      let responseContent = '';

      for await (const message of messagesResponse) {
        if (message.role === 'assistant') {
          for (const content of message.content) {
            if (content.type === 'text') {
              responseContent = (content as any).text.value;
              break;
            }
          }
          if (responseContent) break;
        }
      }

      // Cleanup thread
      await this.client.threads.delete(thread.id);

      return {
        success: true,
        message: responseContent,
        toolCalls,
        metadata: {
          runId: run.id,
          iterations: iteration,
        },
      };
    } catch (error) {
      console.error(`[${this.agentName}] Chat error:`, error);
      return {
        success: false,
        message: error instanceof Error ? error.message : 'Unknown error',
        toolCalls,
      };
    }
  }

  /**
   * Streaming chat with SSE support
   */
  async *chatStream(
    messages: ChatMessage[],
    context?: Record<string, any>
  ): AsyncGenerator<StreamEvent> {
    await this.ensureInitialized();

    if (!this.client || !this.agent) {
      yield {
        type: 'error',
        data: { error: 'Agent not initialized' },
      };
      return;
    }

    const toolCalls: AgentResponse['toolCalls'] = [];

    try {
      yield { type: 'status', data: { status: 'creating_thread', message: 'Starting conversation...' } };

      // Create a thread
      const thread = await this.client.threads.create();

      // Add messages to thread
      for (const msg of messages) {
        if (msg.role !== 'system') {
          await this.client.messages.create(
            thread.id,
            msg.role === 'user' ? 'user' : 'assistant',
            msg.content
          );
        }
      }

      // Add context
      if (context) {
        await this.client.messages.create(
          thread.id,
          'user',
          `[CONTEXT]: ${JSON.stringify(context)}`
        );
      }

      yield { type: 'status', data: { status: 'running', message: 'Processing...' } };

      // Run the agent using createAndPoll
      const poller = this.client.runs.createAndPoll(thread.id, this.agent.id);
      let run = await poller.pollUntilDone();

      const maxIterations = 20;
      let iteration = 0;

      while (run.status === 'requires_action' && iteration < maxIterations) {
        yield { type: 'thinking', data: { status: 'thinking', iteration } };

        // Using type assertion due to SDK beta type inconsistencies
        const requiredAction = run.requiredAction as any;
        if (requiredAction?.submitToolOutputs?.toolCalls) {
          // Emit tool calls
          for (const toolCall of requiredAction.submitToolOutputs.toolCalls) {
            yield {
              type: 'tool_call',
              data: {
                tool: toolCall.function.name,
                arguments: JSON.parse(toolCall.function.arguments),
              },
            };
          }

          const toolOutputs = await this.handleToolCalls(
            requiredAction.submitToolOutputs.toolCalls,
            toolCalls
          );

          // Emit tool results
          for (const output of toolOutputs) {
            yield {
              type: 'tool_result',
              data: {
                toolCallId: output.toolCallId,
                result: output.output ? JSON.parse(output.output) : null,
              },
            };
          }

          const submitResponse = this.client.runs.submitToolOutputs(
            thread.id,
            run.id,
            toolOutputs
          ) as any;

          run = await submitResponse.finalRun();
        } else {
          await new Promise(resolve => setTimeout(resolve, 1000));
          run = await this.client.runs.get(thread.id, run.id);
        }

        iteration++;
      }

      if (iteration >= maxIterations) {
        yield { type: 'error', data: { error: 'Maximum iterations exceeded' } };
        return;
      }

      if (run.status === 'failed') {
        yield { type: 'error', data: { error: run.lastError?.message || 'Run failed' } };
        return;
      }

      // Get response
      const messagesResponse = this.client.messages.list(thread.id);
      let responseContent = '';

      for await (const message of messagesResponse) {
        if (message.role === 'assistant') {
          for (const content of message.content) {
            if (content.type === 'text') {
              responseContent = (content as any).text.value;
              break;
            }
          }
          if (responseContent) break;
        }
      }

      // Cleanup
      await this.client.threads.delete(thread.id);

      yield {
        type: 'complete',
        data: {
          response: responseContent,
          metadata: { runId: run.id, iterations: iteration },
          toolCalls,
        },
      };
    } catch (error) {
      yield {
        type: 'error',
        data: {
          error: error instanceof Error ? error.message : 'Unknown error',
          details: error,
        },
      };
    }
  }

  /**
   * Handle tool calls from the agent
   */
  protected async handleToolCalls(
    toolCalls: RequiredToolCall[],
    toolCallLog: AgentResponse['toolCalls']
  ): Promise<ToolOutput[]> {
    const outputs: ToolOutput[] = [];

    for (const toolCall of toolCalls) {
      const tc = toolCall as any;
      const toolName = tc.function.name;
      const args = JSON.parse(tc.function.arguments);

      console.log(`[${this.agentName}] Executing tool: ${toolName}`);

      let result: any;
      const handler = this.toolHandlers.get(toolName);

      if (handler) {
        try {
          result = await handler(args);
        } catch (error) {
          console.error(`[${this.agentName}] Tool error:`, error);
          result = { error: error instanceof Error ? error.message : 'Tool execution failed' };
        }
      } else {
        result = { error: `Unknown tool: ${toolName}` };
      }

      toolCallLog?.push({
        name: toolName,
        arguments: args,
        result,
      });

      outputs.push({
        toolCallId: tc.id,
        output: JSON.stringify(result),
      });
    }

    return outputs;
  }

  /**
   * Format SSE event for streaming responses
   */
  static formatSSE(event: StreamEvent): string {
    return `event: ${event.type}\ndata: ${JSON.stringify(event.data)}\n\n`;
  }

  /**
   * Get agent status
   */
  getStatus(): {
    name: string;
    initialized: boolean;
    agentId: string | null;
  } {
    return {
      name: this.agentName,
      initialized: !!this.agent,
      agentId: this.agent?.id || null,
    };
  }

  /**
   * Delete the agent from Azure
   */
  async delete(): Promise<void> {
    if (this.client && this.agent) {
      await this.client.deleteAgent(this.agent.id);
      this.agent = null;
      console.log(`[${this.agentName}] Agent deleted`);
    }
  }
}

export default BaseAgent;

File: backend-python/app/agents/base_agent.py
"""
Base Agent Class
Abstract base class for all Azure AI Agents
"""

import json
import logging
from abc import ABC, abstractmethod
from typing import Any, AsyncIterator

from azure.ai.agents.aio import AgentsClient
from azure.ai.agents.models import (
    Agent,
    AgentThread,
    ThreadMessage,
    ThreadRun,
    MessageRole,
    RunStatus,
    FunctionTool,
    FunctionToolDefinition,
    FunctionDefinition,
    ToolDefinition,
    RequiredFunctionToolCall,
    SubmitToolOutputsAction,
)
from azure.identity.aio import DefaultAzureCredential

from app.config import get_settings

logger = logging.getLogger(__name__)


class BaseAgent(ABC):
    """
    Abstract base class for Azure AI Agents.
    Provides common functionality for agent initialization, thread management,
    and message handling.
    """

    def __init__(self, name: str, instructions: str, model: str | None = None):
        """
        Initialize the base agent.

        Args:
            name: Agent name (will be prefixed with configured prefix)
            instructions: System instructions for the agent
            model: Model deployment name (defaults to config value)
        """
        self.settings = get_settings()
        self.name = f"{self.settings.azure_ai_agent_name_prefix}{name}"
        self.instructions = instructions
        self.model = model or self.settings.azure_ai_model_deployment_name

        self._client: AgentsClient | None = None
        self._credential: DefaultAzureCredential | None = None
        self._agent: Agent | None = None
        self._threads: dict[str, AgentThread] = {}
        self._initialized = False

    @property
    def agent_id(self) -> str | None:
        """Get the agent ID if initialized."""
        return self._agent.id if self._agent else None

    @property
    def is_initialized(self) -> bool:
        """Check if agent is initialized."""
        return self._initialized and self._agent is not None

    @abstractmethod
    def get_tools(self) -> list[ToolDefinition]:
        """
        Get tool definitions for this agent.
        Must be implemented by subclasses.
        """
        pass

    @abstractmethod
    async def handle_tool_call(self, tool_name: str, arguments: dict[str, Any]) -> str:
        """
        Handle a tool call from the agent.
        Must be implemented by subclasses.

        Args:
            tool_name: Name of the tool being called
            arguments: Arguments passed to the tool

        Returns:
            Tool output as a string
        """
        pass

    async def initialize(self) -> bool:
        """
        Initialize the agent with Azure AI.
        First searches for existing agent by name, creates new one if not found.

        Returns:
            True if initialization was successful
        """
        try:
            if not self.settings.azure_ai_project_endpoint:
                logger.warning(f"[{self.name}] Azure AI endpoint not configured")
                return False

            logger.info(f"[{self.name}] Initializing agent...")
            logger.info(f"[{self.name}] Using Azure AI endpoint: {self.settings.azure_ai_project_endpoint[:50]}...")

            # Create credential and client
            self._credential = DefaultAzureCredential()
            self._client = AgentsClient(
                endpoint=self.settings.azure_ai_project_endpoint,
                credential=self._credential,
            )
            logger.info(f"[{self.name}] Azure client created successfully")

            # Search for existing agent by name
            logger.info(f"[{self.name}] Searching for existing agent with name: {self.name}")
            existing_agent = None

            try:
                agents_list = self._client.list_agents()
                async for agent in agents_list:
                    if agent.name == self.name:
                        existing_agent = agent
                        logger.info(f"[{self.name}] Found existing agent: {agent.id}")
                        break
            except Exception as list_error:
                logger.warning(f"[{self.name}] Error listing agents: {list_error}")

            if existing_agent:
                # Reuse existing agent
                self._agent = existing_agent
                logger.info(f"[{self.name}] Reusing existing Azure AI Agent: {self._agent.id}")
            else:
                # Create new agent
                logger.info(f"[{self.name}] No existing agent found, creating new agent...")

                # Get tool definitions
                tools = self.get_tools()
                logger.info(f"[{self.name}] Agent will have {len(tools)} tools")

                self._agent = await self._client.create_agent(
                    model=self.model,
                    name=self.name,
                    instructions=self.instructions,
                    tools=tools,
                )
                logger.info(f"[{self.name}] Created new Azure AI Agent: {self._agent.id}")

            self._initialized = True
            logger.info(f"[{self.name}] Agent initialization complete - ID: {self._agent.id}")
            return True

        except Exception as e:
            logger.error(f"[{self.name}] Failed to initialize agent: {e}")
            import traceback
            logger.error(f"[{self.name}] Traceback: {traceback.format_exc()}")
            self._initialized = False
            return False

    async def create_thread(self, session_id: str | None = None) -> str:
        """
        Create a new conversation thread.

        Args:
            session_id: Optional session ID to associate with thread

        Returns:
            Thread ID
        """
        if not self._client or not self._agent:
            raise RuntimeError(f"[{self.name}] Agent not initialized")

        thread = await self._client.threads.create()
        thread_id = thread.id

        if session_id:
            self._threads[session_id] = thread

        logger.debug(f"[{self.name}] Created thread: {thread_id}")
        return thread_id

    async def get_or_create_thread(self, session_id: str) -> str:
        """
        Get existing thread for session or create new one.

        Args:
            session_id: Session identifier

        Returns:
            Thread ID
        """
        if session_id in self._threads:
            return self._threads[session_id].id
        return await self.create_thread(session_id)

    async def send_message(
        self,
        thread_id: str,
        content: str,
        context: dict[str, Any] | None = None,
    ) -> str:
        """
        Send a message to the agent and get a response.

        Args:
            thread_id: Thread ID to send message to
            content: Message content
            context: Optional context data for tool calls

        Returns:
            Agent response text
        """
        if not self._client or not self._agent:
            raise RuntimeError(f"[{self.name}] Agent not initialized")

        # Store context for tool calls
        self._current_context = context or {}

        # Create message
        await self._client.messages.create(
            thread_id=thread_id,
            role=MessageRole.USER,
            content=content,
        )

        # Run agent and process tool calls
        run = await self._client.runs.create(
            thread_id=thread_id,
            agent_id=self._agent.id,
        )

        # Process run until completion
        response = await self._process_run(thread_id, run)

        return response

    async def _process_run(self, thread_id: str, run: ThreadRun) -> str:
        """
        Process a run until completion, handling any tool calls.

        Args:
            thread_id: Thread ID
            run: Initial run object

        Returns:
            Final agent response text
        """
        if not self._client:
            raise RuntimeError(f"[{self.name}] Agent not initialized")

        while run.status in [RunStatus.QUEUED, RunStatus.IN_PROGRESS, RunStatus.REQUIRES_ACTION]:
            if run.status == RunStatus.REQUIRES_ACTION:
                # Handle tool calls
                if isinstance(run.required_action, SubmitToolOutputsAction):
                    tool_outputs = []
                    for tool_call in run.required_action.submit_tool_outputs.tool_calls:
                        if isinstance(tool_call, RequiredFunctionToolCall):
                            try:
                                arguments = json.loads(tool_call.function.arguments)
                                output = await self.handle_tool_call(
                                    tool_call.function.name,
                                    arguments,
                                )
                                tool_outputs.append({
                                    "tool_call_id": tool_call.id,
                                    "output": output,
                                })
                            except Exception as e:
                                logger.error(f"[{self.name}] Tool call error: {e}")
                                tool_outputs.append({
                                    "tool_call_id": tool_call.id,
                                    "output": json.dumps({"error": str(e)}),
                                })

                    # Submit tool outputs
                    run = await self._client.runs.submit_tool_outputs(
                        thread_id=thread_id,
                        run_id=run.id,
                        tool_outputs=tool_outputs,
                    )
            else:
                # Wait and poll for status update
                import asyncio
                await asyncio.sleep(0.5)
                run = await self._client.runs.get(thread_id=thread_id, run_id=run.id)

        if run.status == RunStatus.FAILED:
            error_msg = run.last_error.message if run.last_error else "Unknown error"
            logger.error(f"[{self.name}] Run failed: {error_msg}")
            raise RuntimeError(f"Agent run failed: {error_msg}")

        # Get the latest assistant message
        messages = await self._client.messages.list(thread_id=thread_id)
        for msg in messages.data:
            if msg.role == MessageRole.ASSISTANT:
                # Return first text content
                for content in msg.content:
                    if hasattr(content, "text") and content.text:
                        return content.text.value
                break

        return ""

    async def send_message_streaming(
        self,
        thread_id: str,
        content: str,
        context: dict[str, Any] | None = None,
    ) -> AsyncIterator[str]:
        """
        Send a message and stream the response.

        Args:
            thread_id: Thread ID to send message to
            content: Message content
            context: Optional context data for tool calls

        Yields:
            Response text chunks
        """
        if not self._client or not self._agent:
            raise RuntimeError(f"[{self.name}] Agent not initialized")

        # Store context for tool calls
        self._current_context = context or {}

        # Create message
        await self._client.messages.create(
            thread_id=thread_id,
            role=MessageRole.USER,
            content=content,
        )

        # Create and stream run
        async with await self._client.runs.stream(
            thread_id=thread_id,
            agent_id=self._agent.id,
        ) as stream:
            async for event in stream:
                if hasattr(event, "data") and hasattr(event.data, "delta"):
                    delta = event.data.delta
                    if hasattr(delta, "content"):
                        for content_block in delta.content:
                            if hasattr(content_block, "text") and content_block.text:
                                yield content_block.text.value

    async def cleanup(self) -> None:
        """Clean up agent resources."""
        try:
            if self._client and self._agent:
                await self._client.delete_agent(self._agent.id)
                logger.info(f"[{self.name}] Agent deleted: {self._agent.id}")

            # Clear threads
            self._threads.clear()
            self._agent = None
            self._initialized = False

            # Close credential
            if self._credential:
                await self._credential.close()
                self._credential = None

            # Close client (if it has close method)
            if self._client and hasattr(self._client, "close"):
                await self._client.close()
            self._client = None

        except Exception as e:
            logger.error(f"[{self.name}] Error during cleanup: {e}")

    def create_function_tool(
        self,
        name: str,
        description: str,
        parameters: dict[str, Any],
    ) -> FunctionToolDefinition:
        """
        Helper to create a function tool definition.

        Args:
            name: Function name
            description: Function description
            parameters: JSON Schema for parameters

        Returns:
            FunctionToolDefinition
        """
        return FunctionToolDefinition(
            function=FunctionDefinition(
                name=name,
                description=description,
                parameters=parameters,
            )
        )
